src/v/net/conn_quota.cc

```diff
@@ -371,23 +373,28 @@ bool conn_quota::should_leave_reclaim(home_allowance& allowance) {
  * Broadcast (in the background) to all other shards that they
  * may clear the reclaim flag.
  */
-void conn_quota::cancel_reclaim_to(ss::net::inet_address addr) {
+void conn_quota::cancel_reclaim_to(
+  ss::net::inet_address addr, ss::lw_shared_ptr<home_allowance> allowance) {
     assert_on_home(addr);
 
     vlog(rpc::rpclog.trace, "cancel_reclaim_to({})", addr);
 
-    // Guaranteed to have units because of precheck in
-    // should_leave_reclaim
-    auto& allowance = get_home_allowance(addr);
-    auto units = allowance.reclaim_lock.try_get_units().value();
-    allowance.reclaim = false;
-
-    ssx::spawn_with_gate(_gate, [this, addr, units = std::move(units)] {
-        return container().invoke_on_others(
-          [addr](conn_quota& cq) -> ss::future<> {
-              cq.cancel_reclaim_from(addr);
-              return ss::now();
-          });
+    ssx::spawn_with_gate(_gate, [this, allowance = std::move(allowance), addr] {
+        // Re-check conditions are still suitable.
+        if (should_leave_reclaim(*allowance)) {
+            // Guaranteed to have units because of precheck in
+            // should_leave_reclaim
+            auto units = allowance->reclaim_lock.try_get_units().value();
+            allowance->reclaim = false;
+            return container()
+              .invoke_on_others([addr](conn_quota& cq) -> ss::future<> {
+                  cq.cancel_reclaim_from(addr);
+                  return ss::now();
+              })
+              .finally([allowance, u = std::move(units)]() {});
```

Comment:
here's a fun one: if this is the last reference to `allowance` when the lambda is destroyed, the units `u` probably _have_ to be destroyed first since it will try to decrement the units on the semaphore owned by `allowance`? I guess two finally blocks could be used to make the destruction order clear. it's a bit of a weird scenario that I don't recall seeing before.

Reply:
Interesting point!  I've split this up into two `finally` to make it explicitly safe.